Intellij: Task sequence for task 'compileQuarkusGeneratedSourcesJava' not correct (Gradle, Kotlin, Multi-Module Project)

### Describe the bug

We use `Quarkus`, `Kotlin` and `Gradle` in our various projects.

In `multi-module projects` we get the following error in `Intellij`:

```
FAILURE: Build failed with an exception.

* What went wrong:
Some problems were found with the configuration of task ':app-editor-server:compileKotlin' (type 'KotlinCompile').
  - Gradle detected a problem with the following location: '/home/user1/development/sources/app-editor/app-editor-server/build/classes/java/quarkus-generated-sources'.
    
    Reason: Task ':app-editor-server:compileKotlin' uses this output of task ':app-editor-server:compileQuarkusGeneratedSourcesJava' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':app-editor-server:compileQuarkusGeneratedSourcesJava' as an input of ':app-editor-server:compileKotlin'.
      2. Declare an explicit dependency on ':app-editor-server:compileQuarkusGeneratedSourcesJava' from ':app-editor-server:compileKotlin' using Task#dependsOn.
      3. Declare an explicit dependency on ':app-editor-server:compileQuarkusGeneratedSourcesJava' from ':app-editor-server:compileKotlin' using Task#mustRunAfter.
    
    For more information, please refer to https://docs.gradle.org/8.6/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.
  - Gradle detected a problem with the following location: '/home/user1/development/sources/app-editor/app-editor-server/build/classes/java/quarkus-generated-sources/grpc'.
    
    Reason: Task ':app-editor-server:compileKotlin' uses this output of task ':app-editor-server:compileQuarkusGeneratedSourcesJava' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':app-editor-server:compileQuarkusGeneratedSourcesJava' as an input of ':app-editor-server:compileKotlin'.
      2. Declare an explicit dependency on ':app-editor-server:compileQuarkusGeneratedSourcesJava' from ':app-editor-server:compileKotlin' using Task#dependsOn.
      3. Declare an explicit dependency on ':app-editor-server:compileQuarkusGeneratedSourcesJava' from ':app-editor-server:compileKotlin' using Task#mustRunAfter.
    
    For more information, please refer to https://docs.gradle.org/8.6/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.
  - Gradle detected a problem with the following location: '/home/user1/development/sources/app-editor/app-editor-server/build/classes/java/quarkus-generated-sources/avdl'.
    
    Reason: Task ':app-editor-server:compileKotlin' uses this output of task ':app-editor-server:compileQuarkusGeneratedSourcesJava' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':app-editor-server:compileQuarkusGeneratedSourcesJava' as an input of ':app-editor-server:compileKotlin'.
      2. Declare an explicit dependency on ':app-editor-server:compileQuarkusGeneratedSourcesJava' from ':app-editor-server:compileKotlin' using Task#dependsOn.
      3. Declare an explicit dependency on ':app-editor-server:compileQuarkusGeneratedSourcesJava' from ':app-editor-server:compileKotlin' using Task#mustRunAfter.
    
    For more information, please refer to https://docs.gradle.org/8.6/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.
  - Gradle detected a problem with the following location: '/home/user1/development/sources/app-editor/app-editor-server/build/classes/java/quarkus-generated-sources/avpr'.
    
    Reason: Task ':app-editor-server:compileKotlin' uses this output of task ':app-editor-server:compileQuarkusGeneratedSourcesJava' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':app-editor-server:compileQuarkusGeneratedSourcesJava' as an input of ':app-editor-server:compileKotlin'.
      2. Declare an explicit dependency on ':app-editor-server:compileQuarkusGeneratedSourcesJava' from ':app-editor-server:compileKotlin' using Task#dependsOn.
      3. Declare an explicit dependency on ':app-editor-server:compileQuarkusGeneratedSourcesJava' from ':app-editor-server:compileKotlin' using Task#mustRunAfter.
    
    For more information, please refer to https://docs.gradle.org/8.6/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.
  - Gradle detected a problem with the following location: '/home/user1/development/sources/app-editor/app-editor-server/build/classes/java/quarkus-generated-sources/avsc'.
    
    Reason: Task ':app-editor-server:compileKotlin' uses this output of task ':app-editor-server:compileQuarkusGeneratedSourcesJava' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':app-editor-server:compileQuarkusGeneratedSourcesJava' as an input of ':app-editor-server:compileKotlin'.
      2. Declare an explicit dependency on ':app-editor-server:compileQuarkusGeneratedSourcesJava' from ':app-editor-server:compileKotlin' using Task#dependsOn.
      3. Declare an explicit dependency on ':app-editor-server:compileQuarkusGeneratedSourcesJava' from ':app-editor-server:compileKotlin' using Task#mustRunAfter.
    
    For more information, please refer to https://docs.gradle.org/8.6/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 15s
```

The error only occurs when compiling with the Intellij integrated function `Rebuild Project`. Compiling with `./gradlew build` works without problems.

We were able to determine the following prerequisites for the error to occur in Intellij:
- Multi-Module project
- Kotlin
- Gradle

To workaround the problem we use:

```kotlin
subprojects {
   //...

    tasks.matching { it.name == "compileQuarkusGeneratedSourcesJava" }.configureEach {
        mustRunAfter(tasks.compileKotlin)
    }

   //...
}
```


### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

So far I have not managed to create a small reproducer.

### Output of `uname -a` or `ver`

Linux

### Output of `java -version`

Java 17

### Quarkus version or git rev

Quarkus 3.7.4

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle

### Additional information

_No response_

--- a/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/NamedImpl.java
+++ b/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/NamedImpl.java
@@ -1,0 +1,19 @@
+package io.quarkus.gradle;
+
+import java.util.Objects;
+
+import org.gradle.api.Named;
+
+class NamedImpl implements Named {
+
+    private final String name;
+
+    public NamedImpl(String name) {
+        this.name = Objects.requireNonNull(name);
+    }
+
+    @Override
+    public String getName() {
+        return name;
+    }
+}

--- a/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/QuarkusPlugin.java
+++ b/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/QuarkusPlugin.java
@@ -471,39 +471,69 @@ public class QuarkusPlugin implements Plugin<Project> {
 
                     tasks.named(JavaPlugin.COMPILE_JAVA_TASK_NAME, JavaCompile.class,
                             compileJava -> {
+                                // add the code gen sources
+                                final SourceSet generatedSourceSet = sourceSets
+                                        .getByName(QuarkusGenerateCode.QUARKUS_GENERATED_SOURCES);
+                                addCodeGenSourceDirs(compileJava, generatedSourceSet, quarkusExt);
                                 // quarkusGenerateCode is a dependency
                                 compileJava.dependsOn(quarkusGenerateCode);
                                 // quarkusGenerateCodeDev must run before compileJava in case quarkusDev is the target
                                 compileJava.mustRunAfter(quarkusGenerateCodeDev);
-                                // add the code gen sources
-                                addCodeGenSourceDirs(compileJava,
-                                        sourceSets.getByName(QuarkusGenerateCode.QUARKUS_GENERATED_SOURCES), quarkusExt);
+
+                                // The code below needs to be reviewed. There is no test coverage for it.
+                                // We need to properly configure/use these tasks. For now they aren't actually used
+                                // but w/o the code below, in some cases, Gradle may fail on determining task ordering.
+                                // https://github.com/quarkusio/quarkus/issues/45057 states that this issue happens with
+                                // org.gradle.parallel=true when building from IntelliJ
+                                // Similar code is added for compileKotlin (although for Kotlin these task don't seem to exist).
+                                if (tasks.contains(new NamedImpl(generatedSourceSet.getCompileJavaTaskName()))) {
+                                    compileJava.mustRunAfter(tasks.named(generatedSourceSet.getCompileJavaTaskName()));
+                                }
                             });
                     tasks.named(JavaPlugin.COMPILE_TEST_JAVA_TASK_NAME, JavaCompile.class,
                             compileTestJava -> {
-                                compileTestJava.dependsOn(quarkusGenerateCode);
-                                compileTestJava.dependsOn(quarkusGenerateCodeTests);
+                                // add the code gen test sources
+                                final SourceSet generatedSourceSet = sourceSets
+                                        .getByName(QuarkusGenerateCode.QUARKUS_TEST_GENERATED_SOURCES);
+                                addCodeGenSourceDirs(compileTestJava, generatedSourceSet, quarkusExt);
+                                compileTestJava.dependsOn(quarkusGenerateCode, quarkusGenerateCodeTests);
+                                if (tasks.contains(new NamedImpl(generatedSourceSet.getCompileJavaTaskName()))) {
+                                    compileTestJava.mustRunAfter(tasks.named(generatedSourceSet.getCompileJavaTaskName()));
+                                }
                                 if (project.getGradle().getStartParameter().getTaskNames().contains(QUARKUS_DEV_TASK_NAME)) {
                                     compileTestJava.getOptions().setFailOnError(false);
                                 }
-                                // add the code gen test sources
-                                addCodeGenSourceDirs(compileTestJava,
-                                        sourceSets.getByName(QuarkusGenerateCode.QUARKUS_TEST_GENERATED_SOURCES), quarkusExt);
                             });
                 });
 
         project.getPlugins().withId("org.jetbrains.kotlin.jvm", plugin -> {
             quarkusDev.configure(task -> task.shouldPropagateJavaCompilerArgs(false));
             tasks.named("compileKotlin", task -> {
-                task.mustRunAfter(quarkusGenerateCode);
+                final SourceSet generatedSourceSet = project.getExtensions().getByType(SourceSetContainer.class)
+                        .getByName(QuarkusGenerateCode.QUARKUS_GENERATED_SOURCES);
+                addCodeGenSourceDirs(task, generatedSourceSet, quarkusExt);
+                task.dependsOn(quarkusGenerateCode);
                 task.mustRunAfter(quarkusGenerateCodeDev);
-                addCodeGenSourceDirs(task, project.getExtensions().getByType(SourceSetContainer.class)
-                        .getByName(QuarkusGenerateCode.QUARKUS_GENERATED_SOURCES), quarkusExt);
+                if (tasks.contains(new NamedImpl(generatedSourceSet.getCompileJavaTaskName()))) {
+                    task.mustRunAfter(tasks.named(generatedSourceSet.getCompileJavaTaskName()));
+                }
+                String generatedSourcesCompileTaskName = generatedSourceSet.getCompileTaskName("kotlin");
+                if (tasks.contains(new NamedImpl(generatedSourcesCompileTaskName))) {
+                    task.mustRunAfter(tasks.named(generatedSourcesCompileTaskName));
+                }
             });
             tasks.named("compileTestKotlin", task -> {
+                final SourceSet generatedSourceSet = project.getExtensions().getByType(SourceSetContainer.class)
+                        .getByName(QuarkusGenerateCode.QUARKUS_TEST_GENERATED_SOURCES);
+                addCodeGenSourceDirs(task, generatedSourceSet, quarkusExt);
                 task.dependsOn(quarkusGenerateCodeTests);
-                addCodeGenSourceDirs(task, project.getExtensions().getByType(SourceSetContainer.class)
-                        .getByName(QuarkusGenerateCode.QUARKUS_TEST_GENERATED_SOURCES), quarkusExt);
+                if (tasks.contains(new NamedImpl(generatedSourceSet.getCompileJavaTaskName()))) {
+                    task.mustRunAfter(tasks.named(generatedSourceSet.getCompileJavaTaskName()));
+                }
+                final String generatedSourcesCompileTaskName = generatedSourceSet.getCompileTaskName("kotlin");
+                if (tasks.contains(new NamedImpl(generatedSourcesCompileTaskName))) {
+                    task.mustRunAfter(tasks.named(generatedSourcesCompileTaskName));
+                }
             });
         });
     }

--- a/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusGradleUtils.java
+++ b/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusGradleUtils.java
@@ -46,9 +46,9 @@ public class QuarkusGradleUtils {
     }
 
     public static Path mergeClassesDirs(Collection<Path> classesDirs, File tmpDir, boolean populated, boolean test) {
-        List<Path> existingClassesDirs = classesDirs.stream().filter(p -> Files.exists(p)).toList();
+        List<Path> existingClassesDirs = classesDirs.stream().filter(Files::exists).toList();
 
-        if (existingClassesDirs.size() == 0) {
+        if (existingClassesDirs.isEmpty()) {
             return null;
         }
 

--- a/integration-tests/gradle/src/main/resources/grpc-descriptor-set-alternate-output-dir/build.gradle
+++ b/integration-tests/gradle/src/main/resources/grpc-descriptor-set-alternate-output-dir/build.gradle
@@ -16,7 +16,7 @@ repositories {
 
 dependencies {
     implementation enforcedPlatform("${quarkusPlatformGroupId}:${quarkusPlatformArtifactId}:${quarkusPlatformVersion}")
-    implementation 'io.quarkus:quarkus-resteasy'
+    implementation 'io.quarkus:quarkus-rest'
     implementation 'io.quarkus:quarkus-grpc'
 }
 

--- a/integration-tests/gradle/src/main/resources/grpc-descriptor-set-alternate-output/build.gradle
+++ b/integration-tests/gradle/src/main/resources/grpc-descriptor-set-alternate-output/build.gradle
@@ -16,7 +16,7 @@ repositories {
 
 dependencies {
     implementation enforcedPlatform("${quarkusPlatformGroupId}:${quarkusPlatformArtifactId}:${quarkusPlatformVersion}")
-    implementation 'io.quarkus:quarkus-resteasy'
+    implementation 'io.quarkus:quarkus-rest'
     implementation 'io.quarkus:quarkus-grpc'
 }
 

--- a/integration-tests/gradle/src/main/resources/grpc-descriptor-set/build.gradle
+++ b/integration-tests/gradle/src/main/resources/grpc-descriptor-set/build.gradle
@@ -16,7 +16,7 @@ repositories {
 
 dependencies {
     implementation enforcedPlatform("${quarkusPlatformGroupId}:${quarkusPlatformArtifactId}:${quarkusPlatformVersion}")
-    implementation 'io.quarkus:quarkus-resteasy'
+    implementation 'io.quarkus:quarkus-rest'
     implementation 'io.quarkus:quarkus-grpc'
 }
 

--- a/integration-tests/gradle/src/main/resources/grpc-include-project/build.gradle
+++ b/integration-tests/gradle/src/main/resources/grpc-include-project/build.gradle
@@ -16,7 +16,7 @@ repositories {
 
 dependencies {
     implementation enforcedPlatform("${quarkusPlatformGroupId}:${quarkusPlatformArtifactId}:${quarkusPlatformVersion}")
-    implementation 'io.quarkus:quarkus-resteasy'
+    implementation 'io.quarkus:quarkus-rest'
     implementation 'io.quarkus:quarkus-grpc'
 }
 

--- a/integration-tests/gradle/src/main/resources/grpc-multi-module-project/build.gradle
+++ b/integration-tests/gradle/src/main/resources/grpc-multi-module-project/build.gradle
@@ -40,7 +40,7 @@ subprojects {
 
     dependencies {
 
-        implementation 'io.quarkus:quarkus-resteasy'
+        implementation 'io.quarkus:quarkus-rest'
 
         testImplementation 'io.quarkus:quarkus-junit5'
 

